### PR TITLE
Implement VS Code Formatter

### PR DESCRIPTION
**Pull request by theRoughCode**
This PR implements the skeleton for the VS Code formatter.  We have 2 basic rules for now:

1. Strip away any whitespace to the left of `namespace`. For example,
`   namespace Microsoft.Quantum {}` -> `namespace Microsoft.Quantum {}`.
2. Format operation definitions. For example:
`operation   Foo(q:Qubit, n   :Int)   :  Bool` -> `operation Foo (q : Qubit, n : Int) : Bool`

### CODE REVIEW

**Comment by theRoughCode** [comment]
cc: @amrondonp, @swosti99

**Comment by swernli** [comment]
Apologies for the delay on the review; I thought I had hit submit but had it sitting open in a browser instead. Great work so far!

**Comment by swernli** [comment]
Looks like there was a failure in the testing of the VSCode extension. Is it possible something in these changes is showing up as a unit test when it shouldn't?
https://dev.azure.com/ms-quantum-public/Microsoft%20Quantum%20(public)/_build/results?buildId=12399&view=logs&j=aa540828-7369-5cca-9a86-6547727f16ff&t=5c10f11e-357e-5f9b-e70d-e2e6470f089e&l=1055

**Comment by swernli** [comment]
Actually, looks like quite a few commits were added between my original review and now. I'm going to take a look back through the files again. Thanks for adding testing!

File: build/test.ps1 [hunk 1]
```diff
@@ -38,7 +38,17 @@ function Test-One {
     }
 }
 
+function Test-VSCode {
+    Param([string] $path);
+
+    Write-Host "##[info]Testing VSCode Extension inside $path..."    
+    Push-Location (Join-Path $PSScriptRoot $path)
+        npm run unittest
```
Comment (swernli) [inline build/test.ps1:46]: I think this should fix the break in the test pipeline.
```suggestion
        npm test
```

File: build/test.ps1 [hunk 2]
```diff
@@ -38,7 +38,17 @@ function Test-One {
...
+    Pop-Location
+}
```
Comment (swernli) [inline build/test.ps1:48]: You'll need to have something like this to make sure the failed tests get reported as expected.
```suggestion
    if  ($LastExitCode -ne 0) {
        Write-Host "##vso[task.logissue type=error;]Failed to test VSCodeExtension at $path."
        $script:all_ok = $False
    }
}
```

File: build/test.ps1 [hunk 3]
```diff
@@ -38,7 +38,17 @@ function Test-One {
...
+
 Test-One '../QsCompiler.sln'
+Test-VSCode '../src/VSCodeExtension'
```
Comment (swernli) [inline build/test.ps1:51]: To match the behavior the build.ps1 uses, you should skip executing VSCodeExtension tests if the environment variable is set to false.
```suggestion
if ($Env:ENABLE_VSIX -ne "false") {
    Test-VSCode '../src/VSCodeExtension'
} else {
    Write-Host "##vso[task.logissue type=warning;]VSIX building skipped due to ENABLE_VSIX variable."
}
```

File: src/VSCodeExtension/src/tests/unit/control-structure.test.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    const code = "if(1 == 1){H(qs[0]);}";
+    const expectedCode = "if(1 == 1){H(qs[0]);}";
+
+    assert.equal(formatter(code, []), expectedCode);
+  });
+
+  describe("space after if rule", () => {
+    it("adds space after if statement", () => {
+      const code = "if(1 == 1){H(qs[0]);}";
+      const expectedCode = "if (1 == 1){H(qs[0]);}";
```
Comment (swernli) [inline src/VSCodeExtension/src/tests/unit/control-structure.test.ts:17]: Some of these test cases will need to be updated once other rules are implemented (such as requiring a space before and after open curly brace, ie `if (1 == 1) { H(qs[0]); }`. But that's ok to leave this as-is for now and expect those updates to be done by later rule authors.

File: src/VSCodeExtension/src/formatter/operation-rules.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Formats operation declarations.
+ * 
+ * @param code incoming code
+ * 
+ * @example `operation   Foo(q:Qubit, n   :Int)   :  Bool` ->
+ * `operation Foo (q : Qubit, n : Int) : Bool`
+ */
+const argsRule = (code: string): string => {
+    const operationMatcher: RegExp = /operation\s*(\w+)\s*\((.*)\)\s*:\s*(\w+)/g;
```
[the remark this answers is not in the document]
Comment (theRoughCode) [reply src/VSCodeExtension/src/formatter/operation-rules.ts]: What's a generic umbrella term for operations and functions?

File: build/test.ps1 (hunk 1 above)
Comment (theRoughCode) [reply build/test.ps1:46]: I think this is failing because I forgot to edit the template `package.json`.

File: src/VSCodeExtension/src/tests/unit/control-structure.test.ts (hunk 4 above)
Comment (theRoughCode) [reply src/VSCodeExtension/src/tests/unit/control-structure.test.ts:17]: cc: @amrondonp

File: build/test.ps1 (hunk 1 above)
Comment (swernli) [reply build/test.ps1:46]: Ah, I see the corresponding change below that adds `unittest` as one of the targets.

File: src/VSCodeExtension/src/formatter/operation-rules.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (swernli) [reply src/VSCodeExtension/src/formatter/operation-rules.ts]: That's a good question, and I don't think there really is one. But if the intent of the file is for checking operation and function declarations, maybe a decent name would be declaration-rules.ts? Then it could also handle user-defined type declarations (`newtype` as documented [here](https://docs.microsoft.com/en-us/quantum/user-guide/language/types#user-defined-types)). That file name change can be made in following PRs, no need to gate this PR on the name change.

File: src/VSCodeExtension/src/formatter/format-document.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+): vscode.TextEdit[] => {
+    const lines: vscode.TextLine[] = Array.from(
+        Array(document.lineCount),
+        (_, lineNo) => document.lineAt(lineNo)
+    );
+
+    const formattedLines: vscode.TextEdit[] = lines
+        .map((line) => {
+            const formattedLine: string = formatter(line.text, rules);
+            return formattedLine !== line.text
```
Comment (cgranade) [inline src/VSCodeExtension/src/formatter/format-document.ts:27]: Does this constrain all formatting rules to work one line at a time?

File: src/VSCodeExtension/src/formatter/rules/control-structure.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { FormatRule } from "../formatter";
+
+/**
+ * Laves just one space after an if statement
```
Comment (cgranade) [inline src/VSCodeExtension/src/formatter/rules/control-structure.ts:4]:
```suggestion
 * Leaves just one space after an if statement.
```

File: src/VSCodeExtension/src/formatter/rules/control-structure.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * @param code input string
+ *
+ * @example `if                              (1 == 1){H(qs[0]);}` ->
+ * `if (1 == 1){H(qs[0]);}`
+ *
+ * more examples in the unit tests
+ */
+export const spaceAfterIf: FormatRule = (code: string): string => {
+    return code.replace(/(^| +|;|})if[ \n]*/, "$1if ");
```
Comment (cgranade) [inline src/VSCodeExtension/src/formatter/rules/control-structure.ts:14]: Using the `*` operator could make the regex match on zero instances of `[ \n]`, such that variable names starting with `if` could be affected. Do you want `+` here to match [1, ∞) copies of `[ \n]`?
```suggestion
    return code.replace(/(^| +|;|})if[ \n]+/, "$1if ");
```

File: src/VSCodeExtension/src/formatter/rules/indent.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Removes whitespace from left of `namespace`.
+ *
+ * @param code input string
+ *
+ * @example `   namespace   Foo   {}` ->
+ * `namespace Foo {}`
+ */
+export const namespaceRule: FormatRule = (code: string) => {
+    const namespaceMatcher: RegExp = /^\s*namespace\s*(\w+)\s*(\S|\S.*\S)\s*$/g;
```
Comment (cgranade) [inline src/VSCodeExtension/src/formatter/rules/indent.ts:15]: Does `\w+` catch dotted names like `Microsoft.Quantum.Arrays`?

File: src/VSCodeExtension/src/formatter/rules/operation.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/**
+ * Formats operation declarations.
+ *
+ * @param code incoming code
+ *
+ * @example `operation   Foo(q:Qubit,  n   :Int)   :  Bool` ->
+ * `operation Foo (q : Qubit, n : Int) : Bool`
```
Comment (cgranade) [inline src/VSCodeExtension/src/formatter/rules/operation.ts:10]: I'd generally suggest not including a space between the operation name and its inputs.

+@swernli for style guide discussion.

File: src/VSCodeExtension/src/tests/unit/control-structure.test.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
+import "mocha";
```
Comment (cgranade) [inline src/VSCodeExtension/src/tests/unit/control-structure.test.ts:1]: I'd suggest a unit test to ensure that `if` occuring in the beginning, middle, or end of an identifier name doesn't get replaced.

File: src/VSCodeExtension/src/tests/unit/indent.test.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import "mocha";
```
Comment (cgranade) [inline src/VSCodeExtension/src/tests/unit/indent.test.ts:1]: I'd suggest unit tests to make sure dotted names are preserved as namespace names.

File: src/VSCodeExtension/src/formatter/format-document.ts (hunk 6 above)
Comment (theRoughCode) [reply src/VSCodeExtension/src/formatter/format-document.ts:27]: Yup. I'm not sure how to match the whole document at once. Will definitely have to change this to handle multi-line matching.

File: src/VSCodeExtension/src/formatter/rules/operation.ts (hunk 10 above)
Comment (swernli) [reply src/VSCodeExtension/src/formatter/rules/operation.ts:10]: Yup! That's one of the open questions, and actually the next one I think we could tackle.

File: src/VSCodeExtension/src/formatter/rules/control-structure.ts (hunk 8 above)
Comment (theRoughCode) [reply src/VSCodeExtension/src/formatter/rules/control-structure.ts:14]: We also want to match none in the case of `if(...)` -> `if (...)`. I would suggest the following
```suggestion
    return code.replace(/(^| +|;|})if[ |\n]*\(/, "$1if ");
```

cc: @amrondonp who wrote this

File: src/VSCodeExtension/src/formatter/operation-rules.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (cgranade) [reply src/VSCodeExtension/src/formatter/operation-rules.ts]: FYI, we've generally used "callable" to mean "function or operation."

File: src/VSCodeExtension/src/formatter/rules/indent.ts (hunk 9 above)
Comment (theRoughCode) [reply src/VSCodeExtension/src/formatter/rules/indent.ts:15]: Good catch. What do you think of this:

```javascript
/^\s*namespace\s+(\w+(?:\.\w+)*)\s*({|{.*})\s*/$
```

File: src/VSCodeExtension/src/tests/unit/control-structure.test.ts (hunk 11 above)
Comment (theRoughCode) [reply src/VSCodeExtension/src/tests/unit/control-structure.test.ts:1]: cc: @amrondonp

File: src/VSCodeExtension/src/formatter/rules/operation.ts (hunk 10 above)
Comment (theRoughCode) [reply src/VSCodeExtension/src/formatter/rules/operation.ts:10]: Deferring to teams channel for discussion.

File: src/VSCodeExtension/src/tests/unit/control-structure.test.ts (hunk 4 above)
Comment (amrondonp) [reply src/VSCodeExtension/src/tests/unit/control-structure.test.ts]: The intention of this rule is only to handle the space after the if statement, We obviously need integration tests that test several rules at one but this specific rule only is concerned about the spacing after the if

File: src/VSCodeExtension/src/formatter/format-document.ts (hunk 6 above)
Comment (amrondonp) [reply src/VSCodeExtension/src/formatter/format-document.ts:27]: we could just go lines.join(""); counldn't we?

File: src/VSCodeExtension/src/formatter/rules/control-structure.ts (hunk 8 above)
Comment (amrondonp) [reply src/VSCodeExtension/src/formatter/rules/control-structure.ts:14]: Good suggestion, @theRoughCode thanks!

File: src/VSCodeExtension/src/formatter/rules/indent.ts (hunk 9 above)
Comment (amrondonp) [reply src/VSCodeExtension/src/formatter/rules/indent.ts:15]: Please make a test for every edge case you encounter

File: src/VSCodeExtension/src/tests/unit/control-structure.test.ts (hunk 11 above)
Comment (amrondonp) [reply src/VSCodeExtension/src/tests/unit/control-structure.test.ts:1]: Sure, Will do

File: src/VSCodeExtension/src/formatter/format-document.ts (hunk 6 above)
Comment (theRoughCode) [reply src/VSCodeExtension/src/formatter/format-document.ts:27]: It would be a bit more involved than that. We need to return `vscode.TextEdit[]` instead of just the final formatted string.
